plugins/cleaners.cpp

```diff
@@ -120,136 +295,165 @@ command_result cleanitems (color_ostream &out)
+                continue; // Dead or off map
+            auto pos = Units::getPosition(unit);
+            if (!pos.isValid() || !bounds.containsPos(pos))
+                continue;
+        }
+        TRACE(log, out).print("Selected\n");
+
+        if (!unit->body.spatters.empty())
         {
             for (size_t j = 0; j < unit->body.spatters.size(); j++)
                 delete unit->body.spatters[j];
             cleaned_units++;
             cleaned_total += unit->body.spatters.size();
             unit->body.spatters.clear();
+            DEBUG(log, out).print("Cleaned unit #%d\n", unit->id);
         }
     }
-    if (cleaned_total)
+    if (cleaned_total > 0)
         out.print("Removed %d contaminants from %d creatures.\n", cleaned_total, cleaned_units);
     return CR_OK;
 }
 
-command_result cleanplants (color_ostream &out)
-{
-    // Invoked from clean(), already suspended
-    int cleaned_plants = 0, cleaned_total = 0;
-    for (auto plant : world->plants.all)
-    {
-        if (plant->contaminants.size())
-        {
-            for (size_t j = 0; j < plant->contaminants.size(); j++)
-                delete plant->contaminants[j];
-            cleaned_plants++;
-            cleaned_total += plant->contaminants.size();
-            plant->contaminants.clear();
-        }
-    }
-    if (cleaned_total)
-        out.print("Removed %d contaminants from %d plants.\n", cleaned_total, cleaned_plants);
-    return CR_OK;
-}
-
-command_result spotclean (color_ostream &out, vector <string> & parameters)
-{
-    // HOTKEY COMMAND: CORE ALREADY SUSPENDED
-    if (cursor->x < 0)
-    {
-        out.printerr("The cursor is not active.\n");
-        return CR_WRONG_USAGE;
-    }
+command_result spotclean(color_ostream &out, vector<string> &parameters)
+{   // Hotkey command, already suspended
+    DEBUG(log, out).print("Doing spotclean.\n");
     if (!Maps::IsValid())
     {
         out.printerr("Map is not available.\n");
         return CR_FAILURE;
     }
-    df::map_block *block = Maps::getTileBlock(cursor->x, cursor->y, cursor->z);
-    if (block == NULL)
+    auto pos = Gui::getCursorPos();
+    if (!pos.isValid())
+    {
+        out.printerr("The keyboard cursor is not active.\n");
+        return CR_WRONG_USAGE;
+    }
+    auto block = Maps::getTileBlock(pos);
+    if (!block)
     {
         out.printerr("Invalid map block selected!\n");
         return CR_FAILURE;
     }
+    clean_options options;
+    options.mud = true;
+    options.snow = true;
 
-    for (auto evt : block->block_events)
-    {
-        if (evt->getType() != block_square_event_type::material_spatter)
-            continue;
-        // type verified - recast to subclass
-        df::block_square_event_material_spatterst *spatter = (df::block_square_event_material_spatterst *)evt;
-        clean_mud_safely(spatter, block->map_pos, df::coord(cursor->x % 16, cursor->y % 16, 0));
-    }
+    clean_block(out, block, cuboid(pos), options);
     return CR_OK;
 }
```

Comment:
Could make `spotclean` an alias for `clean here -mds` or possibly `clean here -ads`? (Should we also include tree droppings? A full command can be used if you want to avoid cleaning those.)
I also don't think we can assume core is suspended unless invoking it from the `show` console ensures that.